src/com/amazon/ionelement/impl/AnyElementBase.kt

```diff
@@ -0,0 +1,233 @@
+    final override fun asSymbolOrNull(): SymbolElement? = requireTypeAndCastOrNull(SYMBOL)
+    final override fun asTimestamp(): TimestampElement = requireTypeAndCast(TIMESTAMP)
+    final override fun asTimestampOrNull(): TimestampElement? = requireTypeAndCastOrNull(TIMESTAMP)
+    final override fun asLob(): LobElement  = requireTypeAndCast(BLOB, CLOB)
+    final override fun asLobOrNull(): LobElement? = requireTypeAndCastOrNull(BLOB, CLOB)
+    final override fun asBlob(): BlobElement = requireTypeAndCast(BLOB)
+    final override fun asBlobOrNull(): BlobElement? = requireTypeAndCastOrNull(BLOB)
+    final override fun asClob(): ClobElement = requireTypeAndCast(CLOB)
+    final override fun asClobOrNull(): ClobElement? = requireTypeAndCastOrNull(CLOB)
+    final override fun asContainer(): ContainerElement = requireTypeAndCast(LIST, STRUCT, SEXP)
+    final override fun asContainerOrNull(): ContainerElement? = requireTypeAndCastOrNull(LIST, STRUCT, SEXP)
+    final override fun asSeq(): SeqElement = requireTypeAndCast(LIST, SEXP)
+    final override fun asSeqOrNull(): SeqElement? = requireTypeAndCastOrNull(LIST, SEXP)
+    final override fun asList(): ListElement = requireTypeAndCast(LIST)
+    final override fun asListOrNull(): ListElement? = requireTypeAndCastOrNull(LIST)
+    final override fun asSexp(): SexpElement = requireTypeAndCast(SEXP)
+    final override fun asSexpOrNull(): SexpElement? = requireTypeAndCastOrNull(SEXP)
+    final override fun asStruct(): StructElement = requireTypeAndCast(STRUCT)
+    final override fun asStructOrNull(): StructElement? = requireTypeAndCastOrNull(STRUCT)
+
+    // These are overridden in the narrow implementations of [IonElement].
+    // The default implementations here throw, complaining about an unexpected type.
+    override val booleanValue: Boolean get() = errIfNotTyped(BOOL)
+    override val longValue: Long get() = errIfNotTyped(INT)
+    override val bigIntegerValue: BigInteger get() = errIfNotTyped(INT)
+    override val textValue: String get() = errIfNotTyped(STRING, SYMBOL)
+    override val stringValue: String get() = errIfNotTyped(STRING)
+    override val symbolValue: String get() = errIfNotTyped(SYMBOL)
+    override val decimalValue: Decimal get() = errIfNotTyped(DECIMAL)
+    override val doubleValue: Double get() = errIfNotTyped(FLOAT)
+    override val timestampValue: Timestamp get() = errIfNotTyped(TIMESTAMP)
+    override val bytesValue: IonByteArray get() = errIfNotTyped(BLOB, CLOB)
+    override val blobValue: IonByteArray get() = errIfNotTyped(BLOB)
+    override val clobValue: IonByteArray get() = errIfNotTyped(CLOB)
+    override val containerValues: Iterable<AnyElement> get() = errIfNotTyped(LIST, SEXP, STRUCT)
+    override val seqValues: Iterable<AnyElement> get() = errIfNotTyped(LIST, SEXP)
+    override val listValues: Iterable<AnyElement> get() = errIfNotTyped(LIST)
+    override val sexpValues: Iterable<AnyElement> get() = errIfNotTyped(SEXP)
```

Comment:
Do we really need all of these variants?  I thought the original idea was to force a narrowing for the `AnyElement` case.  I mean, I see it trying to be consistent with the `String`/`Symbol` dichotomy, but in practice I wonder if this is worthwhile.

Reply:
Also, `size` is not available (or maybe I missed it).  What about  the other accessors too?  I think that was part of the rationale to keep the value-ness in the narrowed type.

Reply:
Well, if we're narrowing the types then these should be:

    override val containerValues: Collection<AnyElement> get() = errIfNotTyped(LIST, SEXP, STRUCT)
    override val seqValues: List<AnyElement> get() = errIfNotTyped(LIST, SEXP)
    override val listValues: List<AnyElement> get() = errIfNotTyped(LIST)
    override val sexpValues: List<AnyElement> get() = errIfNotTyped(SEXP)

Reply:
Next revision has this.